### ISSUE TITLE
Parallelization idiom for per subdataset processing

### ISSUE THREAD

**Issue by mih** [datalad/datalad#4419]
Scenario:

```
datalad create-sibling-ria -d . --recursive 3258.85s user 1823.04s system 104% cpu 1:21:14.76 total
```

This is an example of an operation that is done completely independently across datasets, hence a perfect candidate for parallelization (much simpler than `save` or `get`), and can be found elsewhere too (e.g. `push`). They all discover datasets to process first and then process them. It would be instrumental to identify THE paradigm how to deal with that. The potential for speed-up is massive. The above case is equivalent to creating 4300 Git repositories in disjoint folders (no hierarchies). This could be much much faster than 90 min.

**Comment by yarikoptic** [comment]
> They all discover datasets to process first and then process them.

if discovery result is a generator, then it could become even faster in cases where discovery takes awhile and also IO bound in the main process... IMHO something to keep in mind.

**Comment by mih** [comment]
While DataLad itself isn't yet capable of doing this nicely on its own, one can feel that it is unix-friendly. Let's say we have a dataset hierarchy with 1000 immediate subdatasets, where each has even more subdatasets. The task is to publish them quickly. We can simply use `xargs` to achieve parallelization:

```
% datalad -f '{path}' subdatasets | xargs -n 1 -P 10 datalad push -r --to target -d
```

This uses `subdatasets` to discover all first level subdatasets (very fast), and then hands over the list of paths to all 1000 of them to `xargs`, which makes one call to `datalad push` per path (`-n 1`), but performs 10 of them in parallel (`-P 10`). This works, because `push -d ...` is sorted last, so an individual push for each given dataset is made. Of course this is only "legal", because I know that this is an operation that is independent across datasets.

I also could have used `subdatasets -r` (instead of `push -r`) to fully resolve the list of datasets to push, but that would have been slower, due to the reporting order. With the setup above the (slower) recursion part is already done in parallel.

@adswa This might be something for the handbook.

**Comment by mih** [comment]
This information will make it into the handbook via https://github.com/datalad-handbook/book/pull/445

There is no need for keeping this open.

**Comment by mih** [comment]
FTR:

`% datalad -f '{path}' subdatasets | xargs -n 1 -P 10 datalad get -d`

works just as nice for cross-dataset download parallelization.

**Comment by mih** [comment]
FTR:

`datalad-run`-based, within-dataset, parallel compression of all `.nii` files
```
% datalad run 'find . -name "*.nii" -a ! -wholename "*.git/annex/*" | xargs -n 1 -P 10 bash -c "gzip -9 -c \$1 > \$1.gz && rm \$1" COMPRESSME'
```
